src/mlpro/bf/systems.py

```diff
@@ -807,4 +811,410 @@ def _compute_broken(self, p_state: State) -> bool:
+                  p_fct_success : FctSuccess = None, 
+                  p_fct_broken : FctBroken = None, 
+                  p_visualize : bool = False, 
+                  p_logging = Log.C_LOG_ALL ):
+
+        SystemBase.__init__( self,
+                             p_mode=p_mode, 
+                             p_latency=p_latency, 
+                             p_fct_strans=p_fct_strans, 
+                             p_fct_success=p_fct_success, 
+                             p_fct_broken=p_fct_broken, 
+                             p_visualize=p_visualize, 
+                             p_logging=p_logging )
+
+        self._controllers       = []
+        self._mapping_actions   = {}
+        self._mapping_states    = {}
+
+    
+## -------------------------------------------------------------------------------------------------
+    def reset(self, p_seed=None) -> None:
+        """
+        Resets the system by calling method SystemBase.reset() and the related custom method _reset().
+        Furthermore, in real operation mode all assigned controllers are reset as well.
+        """
+
+        SystemBase.reset(self, p_seed)
+
+        if self.get_mode() == Mode.C_MODE_REAL:
+            for con in self._controllers: con.reset()
+            self._import_state()
+
+
+## -------------------------------------------------------------------------------------------------
+    def add_controller(self, p_controller : Controller, p_mapping : list) -> bool:
+        """
+        Adds a controller and a related mapping of states and actions to sensors and actuators.
+
+        Parameters
+        ----------
+        p_controller : Controller
+            Controller object to be added.
+        p_mapping : list
+            A list of mapping tuples following the syntax ( [Type = 'S' or 'A'], [Id state/action] [Id sensor/actuator] )
+
+        Returns
+        -------
+        successful : bool
+            True, if controller and related mapping was added successfully. False otherwise.
+        """
+
+        # 1 Check of mapping entries
+        for entry in p_mapping:
+            if entry[0] == 'S':
+                try:
+                    self._state_space.get_dim(p_id=entry[1])
+                except:
+                    raise ParamError('Invalid state component id "' + str(entry[1] + '"'))
+                    
+            elif entry[0] == 'A':
+                try:
+                    self._action_space.get_dim(p_id=entry[1])
+                except:
+                    raise ParamError('Invalid action component id "' + str(entry[1] + '"'))
+
+            else:
+                raise ParamError('Type "' + entry[0] + '" not valid!')
+
+
+        # 2 Takeover of mapping entries and controller
+        for entry in p_mapping:
+            if entry[0] == 'S':
+                self._mapping_states[entry[1]] = ( p_controller, entry[2] )
+            else:
+                self._mapping_actions[entry[1]] = ( p_controller, entry[2] )
```

Comment:
Here I want to ensure that the internal mapping is extended only if all mapping entries are valid...

Reply:
Yes, I understand and it can be simplified with one loop only. Like the following:


```python
for entry in p_mapping:
    if entry[0] == 'S':
        try:
            self._state_space.get_dim(p_id=entry[1])
        except:
            raise ParamError('Invalid state component id "' + str(entry[1] + '"'))
        else:
            self._mapping_states[entry[1]] = ( p_controller, entry[2] )
            
    elif entry[0] == 'A':
        try:
            self._action_space.get_dim(p_id=entry[1])
        except:
            raise ParamError('Invalid action component id "' + str(entry[1] + '"'))
        else:
            self._mapping_actions[entry[1]] = ( p_controller, entry[2] )

    else:
        raise ParamError('Type "' + entry[0] + '" not valid!')
```

Reply:
The entire method is reworked now. But it's still a 2-pass procedure...